### PR TITLE
Dump PAL filemenu section

### PR DESCRIPTION
**Pull request by marijnvdwerf**

### CODE REVIEW

**Comment by BowserSlug** [comment]
🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀 This PR matches 69 functions (+0.43%) on `pal`.

**Review by ethteck — APPROVED** [review]
Added some cleanup, let's wait for alex/clov to look at it 2

File: include/enums.h [hunk 1]
```diff
@@ -6269,41 +6269,47 @@ enum RushFlags {
...
+    /* 24 */ FILE_MESSAGE_OK_TO_COPY_TO_THIS_FILE,     // OK to copy to this file?[End]
+    /* 25 */ FILE_MESSAGE_START_GAME_WITH,             // Start game with[End]
+    /* 26 */ FILE_MESSAGE_FILE_26,                     // File[End]
+    /* 27 */ FILE_MESSAGE_HAS_BEEN_DELETED,            // has been deleted.[End]
+    /* 28 */ FILE_MESSAGE_28,                          // [End]
+    /* 29 */ FILE_MESSAGE_COPY_FROM,                   // Copy from[End]
+    /* 30 */ FILE_MESSAGE_TO,                          // to[End]
+    /* 31 */ FILE_MESSAGE_HAS_BEEN_CREATED,            // has been created.[End]
+#if VERSION_PAL
+    UNK1,
```
Comment (marijnvdwerf) [inline include/enums.h:6307]: `FILE_MESSAGE_ENTER_A_FILE_NAME` is at the right place, with this define, but I don't know where exactly UNK1 and UNK2 should be placed

File: src/filemenu/filemenu_common.c [hunk 2]
```diff
@@ -717,6 +778,104 @@ void filemenu_update_hidden_name_confirm(
...
+        var_a1 = 0;
+    }
+    if (var_a1 < ARRAY_COUNT(D_filemenu_8024EB80)) {
+        *posX -= D_filemenu_8024EB80[var_a1];
+    } else {
+        *posX -= D_filemenu_8024EB80[ARRAY_COUNT(D_filemenu_8024EB80) - 1];
+        window->flags &= ~WINDOW_FLAG_INITIAL_ANIMATION;
+        window->flags |= WINDOW_FLAG_HIDDEN;
+    }
+    if (var_a1 >= 0x10) {
```
Comment (marijnvdwerf) [inline src/filemenu/filemenu_common.c:873]: Do we need to express this in relation to `ARRAY_COUNT(D_filemenu_8024EB80)`?

File: src/filemenu/filemenu_common.c [hunk 3]
```diff
@@ -732,6 +891,11 @@ void filemenu_draw_cursor(MenuPanel* menu, s32 baseX, s32 baseY, s32 width, s32
     }
 }
 
+#if VERSION_PAL
+#define PAGE_4 (3)
```
Comment (marijnvdwerf) [inline src/filemenu/filemenu_common.c:895]: Should probably use the same enum as in `filemenu_main.c`.

File: src/filemenu/filemenu_common.c (hunk 2 above)
Comment (ethteck) [reply src/filemenu/filemenu_common.c:873]: I looked at that, but I don't think I can really tell without more context. it's like array count minus 1 or 2 right?

File: include/enums.h (hunk 1 above)
Comment (nanaian) [reply include/enums.h:6307]: We need to split the PAL messages to figure this out I think

File: src/filemenu/filemenu_common.c (hunk 2 above)
Comment (nanaian) [reply src/filemenu/filemenu_common.c:873]: Best to leave `// TODO: ARRAY_COUNT?` on this one I think

File: src/filemenu/filemenu_main.c [hunk 4]
```diff
@@ -25,6 +25,27 @@ extern HudScript HES_OptionMonoOff;
...
 
+enum {
+    PAGE_0,
+    PAGE_1,
+#if !VERSION_PAL
+    PAGE_2,
+#endif
+    PAGE_3,
+    PAGE_4,
+};
```
Comment (nanaian) [inline src/filemenu/filemenu_main.c:36]: Can you move this to a `filemenu.h`?

File: src/filemenu/filemenu_info.c [hunk 5]
```diff
@@ -10,6 +10,12 @@
 #define CREATE_SUCCESS_NUMBER_X 48
 #endif
 
+#if VERSION_PAL
+#define PAGE_ID (3)
+#else
+#define PAGE_ID (4)
+#endif
```
Comment (nanaian) [inline src/filemenu/filemenu_info.c:17]: Use `filemenu.h`

File: src/filemenu/filemenu_common.c (hunk 3 above)
Comment (nanaian) [reply src/filemenu/filemenu_common.c:895]: Yes

File: src/filemenu/filemenu_createfile.c [hunk 6]
```diff
@@ -76,12 +88,49 @@ void filemenu_update_change_layout(
     }
 }
 
+#if VERSION_PAL
+extern u8 D_filemenu_8025095C[4];
+// TODO: move filemenu_filename
+#define filemenu_filename filemenu_filename
```
Comment (nanaian) [inline src/filemenu/filemenu_createfile.c:94]: ?

File: src/filemenu/filemenu_main.c [hunk 7]
```diff
@@ -355,14 +426,27 @@ void filemenu_draw_contents_option_left(
...
+#else
     if (menu->page != 2) {
         if (menu->col == 0 && menu->row == 2) {
             filemenu_set_cursor_goal_pos(WINDOW_ID_FILEMENU_OPTION_LEFT, baseX, baseY + 8);
         }
         filemenu_draw_message(filemenu_get_menu_message(FILE_MESSAGE_DELETE_FILE), baseX + 8 + OFFSET_WIDTH, baseY + 2, 255, 0, 1);
     }
+#endif
 }
+//#endif
```
Comment (nanaian) [inline src/filemenu/filemenu_main.c:445]:
```suggestion
```

File: src/filemenu/filemenu_main.c [hunk 8]
```diff
@@ -411,7 +499,17 @@ void filemenu_draw_contents_option_right(
...
     }
 }
+#endif
 
+#if VERSION_PAL
+void filemenu_draw_contents_file_info(s32 fileIdx,
+                                      MenuPanel* menu,
+                                      s32 baseX, s32 baseY,
+                                      s32 width, s32 height,
+                                      s32 opacity, s32 darkening
```
Comment (nanaian) [inline src/filemenu/filemenu_main.c:509]:
```suggestion
void filemenu_draw_contents_file_info(
    s32 fileIdx,
    MenuPanel* menu,
    s32 baseX, s32 baseY,
    s32 width, s32 height,
    s32 opacity, s32 darkening
```

File: src/filemenu/filemenu_pal.c [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+        .height = 0,
+        .priority = WINDOW_PRIORITY_0,
+        .fpDrawContents = &filemenu_draw_pal_8024d6a0,
+        .tab = NULL,
+        .parentID = WINDOW_ID_FILEMENU_MAIN,
+        .fpUpdate = { .func=&filemenu_update_show_title },
+        .extraFlags = 0,
+        .style = { .customStyle = &filemenu_windowStyles[3] }
+    },
+
```
Comment (nanaian) [inline src/filemenu/filemenu_pal.c:79]: Why is there an empty line between some of these

File: src/filemenu/filemenu_pal.c [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
+    if (D_filemenu_8024F128 > 40) {
+        D_filemenu_8024F128 = 0;
+    }
+
+    dma_copy(titlemenu_flags_ROM_START + (D_filemenu_8024F128 * 0x800), titlemenu_flags_ROM_END, D_802517E0[D_filemenu_8024F124]);
+}
+
+
+void filemenu_pal_cleanup(MenuPanel* arg0) {
+    int i;
```
Comment (nanaian) [inline src/filemenu/filemenu_pal.c:372]:
```suggestion
    s32 i;
```

**Review by nanaian — CHANGES_REQUESTED** [review]
🕺

File: include/enums.h (hunk 1 above)
Comment (marijnvdwerf) [reply include/enums.h:6308]: No, all filemenu messages are (unfortunately) stored in `filemenu_msg.c`. Issue is that they're in the PM encoding, so they're not easy to read.

File: src/filemenu/filemenu_pal.c [hunk 11]
```diff
@@ -75,9 +75,7 @@ MenuWindowBP D_filemenu_8024F1D8[] = {
         .fpUpdate = { .func=&filemenu_update_show_title },
         .extraFlags = 0,
         .style = { .customStyle = &filemenu_windowStyles[3] }
-    },
-
-    {
+    }, {
```
Comment (nanaian) [inline src/filemenu/filemenu_pal.c:78]: I think the style we've gone for before is this:
```c
{
    // ...
},
{
    // ...
},
```
but this is fine!

**Review by nanaian — APPROVED** [review]

**Comment by marijnvdwerf** [comment]
Can we _please_ get a `.clang-format` config to document all these style things?

File: src/filemenu/filemenu_pal.c (hunk 11 above)
Comment (marijnvdwerf) [reply src/filemenu/filemenu_pal.c]: Oh, I copied the style from the single `MenuWindowBP[]` that does it differently. Fixed.

File: src/filemenu/filemenu_pal.c (hunk 9 above)
Comment (marijnvdwerf) [reply src/filemenu/filemenu_pal.c]: Leftover from it being not an array

File: src/filemenu/filemenu_createfile.c (hunk 6 above)
Comment (marijnvdwerf) [reply src/filemenu/filemenu_createfile.c]: Leftover from a previously wrong address

**Comment by nanaian** [comment]
> Can we _please_ get a `.clang-format` config to document all these style things?

Yes please :pray:

Need to convince @ethteck and figure out what to do about formatting EVT scripts per #1004

**Comment by marijnvdwerf** [comment]
`filemenu_pal.c` seems to be about the language selection screen. It and its functions probably be renamed to something more descriptive before merging.

**Comment by ethteck** [comment]
> `filemenu_pal.c` seems to be about the language selection screen. It and its functions probably be renamed to something more descriptive before merging.

language_select.c ?

**Comment by marijnvdwerf** [comment]
So without the `filemenu_` prefix?

**Comment by ethteck** [comment]
I gusss with the prefix, but really these should all just be in a filemenu subdirectory so they can lose the prefixes. For now just having the prefix is fine though

**Comment by ethteck** [comment]
I was mistaken. They are already in a directory, so I'm pushing a change to just name the new file filemenu_languageselect for now

**Pull request merged**